packages/react-scheduler/src/ReactScheduler.js

```diff
@@ -190,26 +200,77 @@ if (!ExecutionEnvironment.canUseDOM) {
   ): number {
-    // This assumes that we only schedule one callback at a time because that's
-    // how Fiber uses it.
-    scheduledRICCallback = callback;
+    // Handling multiple callbacks:
+    // For now we implement the behavior expected when the callbacks are
+    // serial updates, such that each update relies on the previous ones
+    // having been called before it runs.
+    // So we call anything in the queue before the latest callback
+
+    let previousCallback;
+    let timeoutTimeFromPreviousCallback;
+    if (scheduledCallback !== null) {
+      // If we have previous callback, save it and handle it below
+      timeoutTimeFromPreviousCallback = timeoutTime;
+      previousCallback = scheduledCallback;
+    }
+    // Then set up the next callback, and update timeoutTime
+    scheduledCallback = callback;
     if (options != null && typeof options.timeout === 'number') {
       timeoutTime = now() + options.timeout;
+    } else {
+      timeoutTime = -1;
+    }
+    // If we have previousCallback, call it. This may trigger recursion.
+    if (
+      previousCallback &&
+      typeof timeoutTimeFromPreviousCallback === 'number'
+    ) {
+      const prevCallbackTimeout: number = timeoutTimeFromPreviousCallback;
+      if (isCurrentlyRunningCallback) {
+        // we are inside a recursive call to rIC
+        // add this callback to a pending queue and run after we exit
+        pendingCallbacks.push({
+          pendingCallback: previousCallback,
+          pendingCallbackTimeout: prevCallbackTimeout,
+        });
+      } else {
+        frameDeadlineObject.didTimeout =
+          timeoutTimeFromPreviousCallback !== -1 &&
+          timeoutTimeFromPreviousCallback <= now();
+        isCurrentlyRunningCallback = true;
+        previousCallback(frameDeadlineObject);
+        isCurrentlyRunningCallback = false;
+        while (pendingCallbacks.length) {
+          // the callback recursively called rIC and new callbacks are pending
+          const {
+            pendingCallback,
+            pendingCallbackTimeout,
+          } = pendingCallbacks.shift();
+          // TODO: pull this into helper method
+          frameDeadlineObject.didTimeout =
+            pendingCallbackTimeout !== -1 && pendingCallbackTimeout <= now();
+          isCurrentlyRunningCallback = true;
+          pendingCallback(frameDeadlineObject);
+          isCurrentlyRunningCallback = false;
+        }
+      }
     }
+
+    // finally, after clearing previous callbacks, schedule the latest one
     if (!isAnimationFrameScheduled) {
       // If rAF didn't already schedule one, we need to schedule a frame.
       // TODO: If this rAF doesn't materialize because the browser throttles, we
       // might want to still have setTimeout trigger rIC as a backup to ensure
       // that we keep performing work.
       isAnimationFrameScheduled = true;
-      requestAnimationFrame(animationTick);
+      return requestAnimationFrame(animationTick);
```

Comment:
So, this method is supposed to return a number, and it seemed to me it should return the request id from the rAF call. But I can remove that change if it needs further discussion.

Reply:
Realized this doesn't quite make sense, because sometimes the initial rAF call just schedules a new rAF call, so even if you got the id of the first call, you might not cancel the overall chain of rAF calls.

I'll just always return 0 as a compromise, but really the return value doesn't mean much and I'd rather drop the return value.

It depends on how the 'cIC' method will work once we support serial and deferred callbacks. If it takes the callback id, then returning a number from 'rIC' makes sense.

Returning an id which could be used to cancel the callback might be useful, but it depends how we implement cIC for deferred priority callbacks.

Reply:
The number is meant to be used as an id to cancel callbacks. This still matters for this PR, since it is possible to have multiple callbacks scheduled at a time (`pendingCallbacks`). When I talked to @sebmarkbage about this a few weeks ago, he suggested that instead of using an id — which requires us to store to map of ids to callbacks — we could store the pending callbacks as a doubly-linked list, and the return value would be the item in the list, which the caller treats as opaque. Then you could remove the callback from the list without needing a map. If you want to do that in a follow up, a map is fine for now, but either way we should account for cancellation in this PR.